Sources/Defaults/Defaults+iCloud.swift

```diff
@@ -0,0 +1,540 @@
+	case idle
+	case syncing
+	case completed
+}
+
+extension Defaults {
+	/**
+	Manages `Defaults.Keys` between the locale and remote storage.
+
+	Depending on the storage, `Defaults.Keys` will be represented in different forms due to storage limitations of the remote storage. The remote storage imposes a limitation of 1024 keys. Therefore, we combine the recorded timestamp and data into a single key. Unlike remote storage, local storage does not have this limitation. Therefore, we can create a separate key (with `defaultsSyncKey` suffix) for the timestamp record.
+	*/
+	public final class iCloudSynchronizer {
+		init(remoteStorage: KeyValueStore) {
+			self.remoteStorage = remoteStorage
+			registerNotifications()
+			remoteStorage.synchronize()
+		}
+
+		deinit {
+			removeAll()
+		}
+
+		@TaskLocal static var timestamp: Date?
+
+		private var cancellables: Set<AnyCancellable> = []
+
+		/**
+		Key for recording the synchronization between `NSUbiquitousKeyValueStore` and `UserDefaults`.
+		*/
+		private let defaultsSyncKey = "__DEFAULTS__synchronizeTimestamp"
+
+		/**
+		A remote key value storage.
+		*/
+		private var remoteStorage: KeyValueStore
+
+		/**
+		A FIFO queue used to serialize synchronization on keys.
+		*/
+		private let backgroundQueue = TaskQueue(priority: .background)
+
+		/**
+		A thread-safe `keys` that manage the keys to be synced.
+		*/
+		@Atomic(value: []) private(set) var keys: Set<Defaults.Keys>
+
+		/**
+		A thread-safe synchronization status monitor for `keys`.
+		*/
+		@Atomic(value: []) private var remoteSyncingKeys: Set<Defaults.Keys>
+
+		// TODO: Replace it with async stream when Swift supports custom executors.
+		private lazy var localKeysMonitor: CompositeUserDefaultsAnyKeyObservation = .init { [weak self] observable in
+			guard
+				let self,
+				let suite = observable.suite,
+				let key = self.keys.first(where: { $0.name == observable.key && $0.suite == suite }),
+				// Prevent triggering local observation when syncing from remote.
+				!self.remoteSyncingKeys.contains(key)
+			else {
+				return
+			}
+
+			self.enqueue {
+				self.recordTimestamp(forKey: key, timestamp: Self.timestamp, source: .local)
+				await self.syncKey(forKey: key, .local)
+			}
+		}
+
+		/**
+		Add new key and start to observe its changes.
+		*/
+		func add(_ keys: [Defaults.Keys]) {
+			self.keys.formUnion(keys)
+			self.syncWithoutWaiting(keys)
+			for key in keys {
+				localKeysMonitor.addObserver(key)
+			}
+		}
+
+		/**
+		Remove key and stop the observation.
+		*/
+		func remove(_ keys: [Defaults.Keys]) {
+			self.keys.subtract(keys)
+			for key in keys {
+				localKeysMonitor.removeObserver(key)
+			}
+		}
+
+		/**
+		Remove all sync keys.
+		*/
+		func removeAll() {
+			localKeysMonitor.invalidate()
+			_keys.modify { $0.removeAll() }
+			_remoteSyncingKeys.modify { $0.removeAll() }
+		}
+
+		/**
+		Explicitly synchronizes in-memory keys and values with those stored on disk.
+		*/
+		func synchronize() {
```

Comment:
Yeah, also think this might cause misleading.
The `synchronize` is trying to expose the `NSUbiquitousKeyValueStore.synchronize()`.
I believe we should rename `sync` to something that more clearly conveys its purpose.
Maybe rename it as sth like `commit`?

Reply:
Do we actually need to expose the method? We can just make `syncOnChange` default to true and remove the option. Then this method would be moot too.